odoo/addons/base/models/res_partner.py

```diff
@@ -517,7 +517,14 @@ def write(self, vals):
             self.invalidate_cache(['user_ids'], self._ids)
             for partner in self:
                 if partner.active and partner.user_ids:
-                    raise ValidationError(_('You cannot archive a contact linked to an internal user.'))
+                    if self.env['res.users'].check_access_rights('write', raise_exception=False):
+                        error_msg = _('You cannot archive a contact linked to an active user. '
+                                      'You first need to archive its associated user.')
+                        action_error = partner.open_user()
+                        raise RedirectWarning(error_msg, action_error, _('Go to user'))
```

Comment:
It might be a good idea to also check whether the current user can edit the partner's user(s) specifically. Also to check after "downgrading" from `su`: if `write` is called in sudo mode this will trigger the RedirectWarning even though the user may not be able to see users at all.